core/node/rpc/sync/operation.go

```diff
@@ -0,0 +1,269 @@
+			}
+			return nil, err
+		case <-syncOp.ctx.Done():
+			return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+		}
+	}
+
+	return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+}
+
+func (syncOp *StreamSyncOperation) RemoveStreamFromSync(
+	ctx context.Context,
+	req *connect.Request[RemoveStreamFromSyncRequest],
+) (*connect.Response[RemoveStreamFromSyncResponse], error) {
+	if req.Msg.GetSyncId() != syncOp.SyncID {
+		return nil, RiverError(Err_INVALID_ARGUMENT, "invalid syncId").Tag("syncId", req.Msg.GetSyncId())
+	}
+
+	op := &subCommand{
+		Ctx:         ctx,
+		RmStreamReq: req,
+		ReplyErr:    make(chan error, 1),
+	}
+
+	if syncOp.sendMsg(op) {
+		select {
+		case err := <-op.ReplyErr:
+			if err == nil {
+				return &connect.Response[RemoveStreamFromSyncResponse]{}, nil
+			}
+			return nil, err
+		case <-syncOp.ctx.Done():
+			return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+		}
+	}
+
+	return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+}
+
+func (syncOp *StreamSyncOperation) CancelSync(
+	_ context.Context,
+	req *connect.Request[CancelSyncRequest],
+) (*connect.Response[CancelSyncResponse], error) {
+	if req.Msg.GetSyncId() != syncOp.SyncID {
+		return nil, RiverError(Err_INVALID_ARGUMENT, "invalid syncId").Tag("syncId", req.Msg.GetSyncId())
+	}
+
+	syncOp.cancel()
+
+	return &connect.Response[CancelSyncResponse]{
+		Msg: &CancelSyncResponse{},
+	}, nil
+}
+
+func (syncOp *StreamSyncOperation) PingSync(
+	ctx context.Context,
+	req *connect.Request[PingSyncRequest],
+) (*connect.Response[PingSyncResponse], error) {
+	if req.Msg.GetSyncId() != syncOp.SyncID {
+		return nil, RiverError(Err_INVALID_ARGUMENT, "invalid syncId").Tag("syncId", req.Msg.GetSyncId())
+	}
+
+	op := &subCommand{
+		Ctx:      ctx,
+		PingReq:  req,
+		ReplyErr: make(chan error, 1),
+	}
+
+	if syncOp.sendMsg(op) {
+		select {
+		case err := <-op.ReplyErr:
+			if err == nil {
+				return &connect.Response[PingSyncResponse]{}, nil
+			}
+			return nil, err
+		case <-syncOp.ctx.Done():
+			return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+		}
+	}
+
+	return nil, RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+}
+
+func (syncOp *StreamSyncOperation) debugDropStream(ctx context.Context, streamID shared.StreamId) error {
+	op := &subCommand{
+		Ctx:             ctx,
+		DebugDropStream: streamID,
+		ReplyErr:        make(chan error, 1),
+	}
+
+	if syncOp.sendMsg(op) {
+		select {
+		case err := <-op.ReplyErr:
+			return err
+		case <-syncOp.ctx.Done():
+			return RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+		}
+	}
+
+	return RiverError(Err_CANCELED, "sync operation cancelled").Tags("syncId", syncOp.SyncID)
+}
+
+func (syncOp *StreamSyncOperation) sendMsg(cmd *subCommand) bool {
```

Comment:
while works this way, one possible refactor is changing sendMsg to do both write and reply read (and returning error if context is done) - then there is no need to explicitly do it in multiple functions.